r/R/compute.R

```diff
@@ -306,3 +306,145 @@ cast_options <- function(safe = TRUE, ...) {
+#'   R vectors before being passed to `fun` and the output is automatically
+#'   constructed with the expected output type via [as_arrow_array()].
+#' @param auto_convert Use `TRUE` to convert inputs before passing to `fun`
+#'   and construct an Array of the correct type from the output. Use this
+#'   option to write functions of R objects as opposed to functions of
+#'   Arrow R6 objects.
+#'
+#' @return `NULL`, invisibly
+#' @export
+#'
+#' @examplesIf arrow_with_dataset()
+#' library(dplyr, warn.conflicts = FALSE)
+#'
+#' some_model <- lm(mpg ~ disp + cyl, data = mtcars)
+#' register_scalar_function(
+#'   "mtcars_predict_mpg",
+#'   function(context, disp, cyl) {
+#'     predict(some_model, newdata = data.frame(disp, cyl))
+#'   },
+#'   in_type = schema(disp = float64(), cyl = float64()),
+#'   out_type = float64(),
+#'   auto_convert = TRUE
+#' )
+#'
+#' as_arrow_table(mtcars) %>%
+#'   transmute(mpg, mpg_predicted = mtcars_predict_mpg(disp, cyl)) %>%
+#'   collect() %>%
+#'   head()
+#'
+register_scalar_function <- function(name, fun, in_type, out_type,
+                                     auto_convert = FALSE) {
+  assert_that(is.string(name))
+
+  scalar_function <- arrow_scalar_function(
+    fun,
+    in_type,
+    out_type,
+    auto_convert = auto_convert
+  )
+
+  # register with Arrow C++ function registry (enables its use in
+  # call_function() and Expression$create())
+  RegisterScalarUDF(name, scalar_function)
+
+  # register with dplyr binding (enables its use in mutate(), filter(), etc.)
+  register_binding(
+    name,
+    function(...) build_expr(name, ...),
+    update_cache = TRUE
+  )
+
+  invisible(NULL)
+}
+
+arrow_scalar_function <- function(fun, in_type, out_type, auto_convert = FALSE) {
+  fun <- as_function(fun)
+
+  # Create a small wrapper function that is easier to call from C++.
+  # This wrapper could be implemented in C/C++ to reduce evaluation
+  # overhead and generate prettier backtraces when errors occur
+  # (probably using a similar approach to purrr).
+  if (auto_convert) {
+    wrapper_fun <- function(context, args) {
+      args <- lapply(args, as.vector)
+      result <- do.call(fun, c(list(context), args))
+      as_arrow_array(result, type = context$output_type)
+    }
+  } else {
+    wrapper_fun <- function(context, args) {
+      do.call(fun, c(list(context), args))
+    }
+  }
+
+  if (is.list(in_type)) {
+    in_type <- lapply(in_type, as_scalar_function_in_type)
+  } else {
+    in_type <- list(as_scalar_function_in_type(in_type))
+  }
+
+  if (is.list(out_type)) {
+    out_type <- lapply(out_type, as_scalar_function_out_type)
+  } else {
+    out_type <- list(as_scalar_function_out_type(out_type))
+  }
+
+  out_type <- rep_len(out_type, length(in_type))
+
+  structure(
+    list(
+      wrapper_fun = wrapper_fun,
+      in_type = in_type,
+      out_type = out_type
+    ),
+    class = "arrow_scalar_function"
+  )
+}
+
+as_scalar_function_in_type <- function(x) {
```

Comment:
Any reason these shouldn't be `as_schema.Field` and `as_schema.DataType`, in which case this function goes away?

Reply:
Make a JIRA for that? Seems reasonable/good-first-issue material